firedrake/preconditioners/asm.py

```diff
@@ -66,6 +89,117 @@ def applyTranspose(self, pc, x, y):
+
+            # Set default solver parameters
+            opts = PETSc.Options(asmpc.getOptionsPrefix())
+            if "pc_asm_type" not in opts:
+                opts["pc_asm_type"] = "basic"
+            if "sub_pc_type" not in opts:
+                opts["sub_pc_type"] = "lu"
+            if "sub_pc_factor_shift_type" not in opts:
+                opts["sub_pc_factor_shift_type"] = "nonzero"
+
+            asmpc.setFromOptions()
+        elif backend == 'tinyasm':
+            asmpc = PETSc.PC().create(comm=pc.comm)
+            asmpc.incrementTabLevel(1, parent=pc)
+            asmpc.setOptionsPrefix(self.prefix + "sub_")
+            asmpc.setOperators(*pc.getOperators())
+            try:
+                from tinyasm import _tinyasm as tasm
+            except ImportError:
+                raise ImportError("To use the TinyASM backend, you must install firedrake with TinySM (firedrake-update --tinyasm).")
+            asmpc.setType("tinyasm")
+            tasm.SetASMLocalSubdomains(
+                asmpc, ises,
+                [W.dm.getDefaultSF() for W in V],
+                [W.value_size for W in V],
+                sum(W.value_size*W.dm.getDefaultSection().getStorageSize() for W in V))
+            asmpc.setFromOptions()
+            asmpc.setUp()
+
+        self.asmpc = asmpc
+
+    def get_patches(self, V):
+        mesh = V._mesh
+        mesh_dm = mesh._topology_dm
+
+        # Obtain the topological entities to use to construct the stars
+        depth = PETSc.Options().getInt(self.prefix+"construct_dim", default=0)
+
+        # Accessing .indices causes the allocation of a global array,
+        # so we need to cache these for efficiency
+        V_local_ises_indices = []
+        for (i, W) in enumerate(V):
+            V_local_ises_indices.append(V.dof_dset.local_ises[i].indices)
+
+        # Build index sets for the patches
+        ises = []
+        (start, end) = mesh_dm.getDepthStratum(depth)
+        for seed in range(start, end):
+            # Only build patches over owned DoFs
+            if mesh_dm.getLabelValue("pyop2_ghost", seed) != -1:
+                continue
+
+            # Create point list from mesh DM
+            pt_array, _ = mesh_dm.getTransitiveClosure(seed, useCone=False)
+
+            # Get DoF indices for patch
+            indices = []
+            for (i, W) in enumerate(V):
+                section = W.dm.getDefaultSection()
+                for p in pt_array.tolist():
+                    dof = section.getDof(p)
+                    if dof <= 0:
+                        continue
+                    off = section.getOffset(p)
+                    # Local indices within W
+                    W_indices = numpy.arange(off*W.value_size, W.value_size * (off + dof), dtype='int32')
+                    indices.extend(V_local_ises_indices[i][W_indices])
+            iset = PETSc.IS().createGeneral(indices, comm=COMM_SELF)
+            ises.append(iset)
+
+        return ises
```

Comment:
Does it make sense to replicate all of `ASMPatchPC`?

Reply:
This is building on top of Scott's PR, so the modification to ASMPatchPC is actually much smaller than this diff suggests. However, there was some duplication in the `if backend == 'petscasm`: ... else ... ` part, which I have now reduced.

Reply:
I think there's some fundamental duplication that we haven't really resolved in my `ASMStarPC` PR (and that's mostly my fault for not getting back to it with any serious effort this week...), trying to see a way forward that doesn't negatively impact `ASMLinesmoothPC`.  I'll take a shot at that now.